text/0000-ember-data-return-promise-from-ds-model-save.md

```diff
@@ -0,0 +1,65 @@
+
+    * Accessing properties on the PromiseObject returned from DS.Model.save()
+        * Use deprecateProperty from Ember to deprecate all the properties that the PromiseProxy Mixin provides. The list of properties can be found here: https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/mixins/promise_proxy.js
+        * Deprecate accessing any unknown properties. For instance, someone might try to grab the value of an attr or computed property from a model, e.g. proxy.get('email').
+    * Calling functions that are not available on RSVP.Promises
+        * Not Safe (issue deprecation): .get/.set/anything else really
+        * Safe (do not issue a deprecation): .then, .catch, .finally
+
+The implementation follows the standard feature-flagging process in Ember and Ember Data. The final implementation will look something like:
+
+```
+  save(options) {
+    const savePromise = this._internalModel.save(options).then(() => this);
+    
+    if (flagEnabled("ds-return-promise-from-model-save")) {
+      return savePromise;
+    } else {
+      return DeprecatedPromiseObject.create({
+        promise: savePromise
+      });
+    }
+  },
+```
+
+## How we teach this
+
+We do not believe this requires any update to the Ember curriculum. API documentation does not need to be updated as it is already documented as returning a promise. This behavior of accessing properties through the PromiseProxyMixin specifically for DS.Model.save() is not documented in the guides.
+
+## Drawbacks
+
+For users relying on this behavior, they will have to refactor their code to either use patterns like async/await, ember-concurrency, or ember-promise-helpers. Alternatively, if all they want to access are properties on the model, they can use the model instead.
+
+## Alternatives
+
+The impact of not doing this prevents further changes in Ember Data and requires more effort to maintain compatibility between Ember Data and older versions of Ember.
+
+## Unresolved questions
+
+* It's quite possible Ember Data users use the isSettled, isRejected, isFulfilled, and isPending properties of the  PromiseProxyMixin to display state in their UI. For example, you may have a component that calls record.save(), and uses isPending to show a loading spinner or isRejected to show UI for error states. Should we point them to community alternatives like [ember-promise-helpers](https://github.com/fivetanley/ember-promise-helpers) or [ember-concurrency](http://ember-concurrency.com/docs/introduction/) in the deprecation message for these properties?
```

Comment:
We could also implement something on top of Promise that exposes these flags.

https://github.com/snewcomer/ember-stateful-promise/blob/main/addon/decorators/stateful-function.js

Reply:
I guess their method could be wrapped with something similar to this or ember-concurrency as well.  Almost every `save` we deal with uses derived state to help the UI along and prevent user initiated bugs. 

So I take that back.  Returning a simple Promise is better.